frontend/apps/reader/modules/readerhighlight.lua

```diff
@@ -189,18 +189,33 @@ function ReaderHighlight:onTapPageSavedHighlight(ges)
 end
 
 function ReaderHighlight:onTapXPointerSavedHighlight(ges)
+    local cur_page
+    -- In scroll mode, we'll need to check for highlights in previous or next
+    -- page too as some parts of them may be displayed
+    local neigbour_pages = self.view.view_mode ~= "page" and 1 or 0
     local pos = self.view:screenToPageTransform(ges.pos)
     for page, _ in pairs(self.view.highlight.saved) do
         local items = self.view.highlight.saved[page]
         if items then
             for i = 1, #items do
+                if not cur_page then
+                    cur_page = self.ui.document:getPageFromXPointer(self.ui.document:getXPointer())
+                end
                 local pos0, pos1 = items[i].pos0, items[i].pos1
-                local boxes = self.ui.document:getScreenBoxesFromPositions(pos0, pos1)
-                if boxes then
-                    for index, box in pairs(boxes) do
-                        if inside_box(pos, box) then
-                            logger.dbg("Tap on hightlight")
-                            return self:onShowHighlightDialog(page, i)
+                -- document:getScreenBoxesFromPositions() is expensive, so we
+                -- first check this item is on current page
+                local page0 = self.ui.document:getPageFromXPointer(pos0)
+                local page1 = self.ui.document:getPageFromXPointer(pos1)
+                local start_page = math.min(page0, page1)
+                local end_page = math.max(page0, page1)
+                if start_page <= cur_page + neigbour_pages and end_page >= cur_page - neigbour_pages then
```

Comment:
Are you sure that `+` and  `-` are the right way around?

NB This is all untested but this is how I'd intuitively write it myself unless I'm misunderstanding what's supposed to happen. At least to me it seems clearer to say "this is how we're adjusting the start page" than "let's pretend the current page is a slightly different one."

```lua
start_page - neigbour_pages <= cur_page and end_page + neigbour_pages >= cur_page
```

Reply:
Yes, I'm sure, and it also feels a bit strange to me.

We want to check the highlight start_page in on the current_page (in page mode), so it could be start_page == cur_page.
In scroll mode, we may be displaying a part of cur_page AND some parts of cur_page -1 OR cur_page +1 : so, we want the highlight start_page to be on cur_page -1, cur_page or cur_page+1 , so : start_page <= cur_page +1.
Same thinking with highlight's end_page, and we got : end_page >= cur_page -1.
It's really cur_page we are adjusting. I can't thing of an alternative writing.

Reply:
And if some big highlight starts on cur_page - 23 and ends on cur_page + 17, it is a highlight to consider.